server/schedule/region_scatterer.go

```diff
@@ -457,16 +457,22 @@ func (r *RegionScatterer) selectStore(group string, peer *metapb.Peer, sourceSto
 }
 
 // selectAvailableLeaderStore select the target leader store from the candidates. The candidates would be collected by
-// the existed peers store depended on the leader counts in the group level.
-func (r *RegionScatterer) selectAvailableLeaderStore(group string, peers map[uint64]*metapb.Peer, context engineContext) uint64 {
+// the existed peers store depended on the leader counts in the group level. Please use this func before scatter spacial engines.
+func (r *RegionScatterer) selectAvailableLeaderStore(group string, region *core.RegionInfo, peers map[uint64]*metapb.Peer, context engineContext) uint64 {
+	sourceStore := r.cluster.GetStore(region.GetLeader().GetStoreId())
+	if sourceStore == nil {
+		log.Error("failed to get the store", zap.Uint64("store-id", region.GetLeader().GetStoreId()), errs.ZapError(errs.ErrGetSourceStore))
+		return 0
+	}
 	leaderCandidateStores := make([]uint64, 0)
+	// use PlacementLeaderSafeguard for con follower and learner in rule
+	filter := filter.NewPlacementLeaderSafeguard(r.name, r.cluster.GetOpts(), r.cluster.GetBasicCluster(), r.cluster.GetRuleManager(), region, sourceStore, false /*allowMoveLeader*/)
 	for storeID := range peers {
 		store := r.cluster.GetStore(storeID)
 		if store == nil {
 			return 0
 		}
-		engine := store.GetLabelValue(core.EngineKey)
-		if len(engine) < 1 {
+		if filter == nil || filter.Target(r.cluster.GetOpts(), store).IsOK() {
```

Comment:
why check the filter is nil?

Reply:
because if placementrule is disabled, filter will be nil.